sidechain/consensus/aura/src/slot_proposer.rs

```diff
@@ -16,59 +16,113 @@
+use log::*;
+use sgx_externalities::SgxExternalitiesTrait;
 use sp_runtime::{
 	traits::{Block, NumberFor},
 	MultiSignature,
 };
 use std::{marker::PhantomData, string::ToString, sync::Arc, time::Duration};
 
+pub type ExternalitiesFor<T> = <T as StfTimedCallsBatchTentativeExecution>::Externalities;
 ///! `SlotProposer` instance that has access to everything needed to propose a sidechain block.
-pub struct SlotProposer<PB: Block, SB: SignedBlock, TopPoolExecutor, BlockComposer> {
+pub struct SlotProposer<PB: Block, SB: SignedBlock, TopPoolExecutor, StfExecutor, BlockComposer> {
 	pub(crate) top_pool_executor: Arc<TopPoolExecutor>,
+	pub(crate) stf_executor: Arc<StfExecutor>,
 	pub(crate) block_composer: Arc<BlockComposer>,
 	pub(crate) parentchain_header: PB::Header,
 	pub(crate) shard: ShardIdentifierFor<SB>,
 	pub(crate) _phantom: PhantomData<PB>,
 }
 
-impl<PB, SB, TopPoolExecutor, BlockComposer> Proposer<PB, SB>
-	for SlotProposer<PB, SB, TopPoolExecutor, BlockComposer>
+impl<PB, SB, TopPoolExecutor, BlockComposer, StfExecutor> Proposer<PB, SB>
+	for SlotProposer<PB, SB, TopPoolExecutor, StfExecutor, BlockComposer>
 where
 	PB: Block<Hash = H256>,
 	NumberFor<PB>: BlockNumberOps,
 	SB: SignedBlock<Public = sp_core::ed25519::Public, Signature = MultiSignature> + 'static,
 	SB::Block: SidechainBlockT<ShardIdentifier = H256, Public = sp_core::ed25519::Public>,
-	TopPoolExecutor: TopPoolCallOperator<ParentchainBlockT = PB> + Send + Sync + 'static,
-	BlockComposer: ComposeBlockAndConfirmation<ParentchainBlockT = PB, SidechainBlockT = SB>
+	StfExecutor: StfTimedCallsBatchTentativeExecution,
+	ExternalitiesFor<StfExecutor>:
+		SgxExternalitiesTrait + SidechainState + SidechainSystemExt + StateHash,
+	TopPoolExecutor: TopPoolCallOperator<PB, SB> + Send + Sync + 'static,
+	BlockComposer: ComposeBlockAndConfirmation<ExternalitiesFor<StfExecutor>, PB, SidechainBlockT = SB>
 		+ Send
 		+ Sync
 		+ 'static,
 {
+	/// Proposes a new sidechain block.
+	///
+	/// This includes the following steps:
+	/// 1) Retrieve all trusted calls from the top pool.
+	/// 2) Calculats the resulting state diff of trusted call batch execution by tentatively
+	///  executing them in a timed batch. No permanent changes state changes in case the block will not be valid.
+	/// 3) Compose the sidechain block and the parentchain confirmation.
 	fn propose(&self, max_duration: Duration) -> Result<Proposal<SB>, ConsensusError> {
-		let latest_onchain_header = &self.parentchain_header;
+		let latest_parentchain_header = &self.parentchain_header;
 
-		let batch_execution_result = self
+		// 1) Retrieve trusted calls from top pool.
+		let trusted_calls = self
 			.top_pool_executor
-			.execute_trusted_calls(latest_onchain_header, &self.shard, max_duration)
+			.get_trusted_calls(&self.shard)
+			.map_err(|e| ConsensusError::Other(e.to_string().into()))?;
+
+		// TODO: remove when we have proper on-boarding of new workers #273.
```

Comment:
Removed & copy pasted from previous `execute_trusted_calls` of the top_pool executor.

Reply:
Yes indeed, @haerdib and I discussed this. The idea is to have this crate be only responsible for executing the trusted getters and thus re-name it accordingly.